imp(precompiles): Add unit test to ensure correct Bech32 addresses for default precompiles

MalteHerrmann opened the pull request:
This PR adds a unit test to ensure we are using the correct Bech32 addresses in the definition of the default precompiles.

This was missed once for the p256 precompile and subsequently fixed in #2550.

GAtom22 commented on precompiles/common/types_test.go:
```diff
@@ -5,13 +5,41 @@ import (
+		sdk.AccAddress(ethcommon.HexToAddress(ics20precompile.PrecompileAddress).Bytes()).String(),
+		sdk.AccAddress(ethcommon.HexToAddress(vestingprecompile.PrecompileAddress).Bytes()).String(),
+		sdk.AccAddress(ethcommon.HexToAddress(bankprecompile.PrecompileAddress).Bytes()).String(),
+	}, common.DefaultPrecompilesBech32)
```
Ideally, we shouldn't have the same address hardcoded twice. There should be only one source of truth. Can we refactor this use smth like `sdk.AccAddress(ethcommon.HexToAddress(bankprecompile.PrecompileAddress).Bytes()).String()` in the definition of the `DefaultPrecompilesBech32 instead of hardcoding these?

MalteHerrmann replied:
I agree that would be the best case, because as it is right now this is also pretty error-prone, because we can just forget to add the address of a new precompile to this list and it will be possible to send funds to it. 

However this would require some bigger refactors it looks like and this PR was just a quick safety measure to at least have a check in place which wasn't there before. The problem I'm seeing why it's not so straight forward is, that the `BlockedAddrs()` method where this slice of Bech32 addresses is passed is necessary to pass into the creation of the bank keeper. Then subsequently, the bank keeper needs to be passed into the creation of the EVM keeper, so we can't just use the `GetAvailablePrecompileAddrs()` method to return the actual addresses.

I can adjust it, so that instead of the hardcoded values we are actually computing the addresses on the spot when running the `BlockedAddrs()` method, but it will still be manually filled with entries, so that won't be solved for now 👀

MalteHerrmann commented:
Hey @GAtom22 I've considered your comments, which I agreed with and adjusted the code in the following ways:

- Removed the `common.DefaultPrecompilesBech32` slice and moved the definition of the blocked precompiles into the `BlockedAddrs` method.
- Added a corresponding test for the blocked precompiles that checks if it's possible to send a transfer to all available precompiles and requires us going forward to specify which should be able to receive funds. This will help to make sure we won't forget to add any to the blocked addresses that we don't mean to ignore there.

However, this has now uncovered that we are not yet accounting for the Ethereum native precompiles as blocked addresses and required some additional checks on the input length for most of our EVM extensions' `RequiredGas` methods.
Please re-review - IMO we should also add the Ethereum native precompiles to the blocked addresses no? 🤔

MalteHerrmann commented on app/app_test.go:
```diff
@@ -45,39 +53,115 @@ func TestEvmosExport(t *testing.T) {
+		ethcommon.HexToAddress("0x0000000000000000000000000000000000000004"),
+		ethcommon.HexToAddress("0x0000000000000000000000000000000000000005"),
+		ethcommon.HexToAddress("0x0000000000000000000000000000000000000006"),
+		ethcommon.HexToAddress("0x0000000000000000000000000000000000000007"),
+		ethcommon.HexToAddress("0x0000000000000000000000000000000000000008"),
```
Interestingly the `0x09` precompile also doesn't work for the simple ETH transfer via an EVM transaction - other than sending via Cosmos transaction to this address.

MalteHerrmann commented on precompiles/bank/bank.go:
```diff
@@ -75,6 +75,10 @@ func (Precompile) Address() common.Address {
 
 // RequiredGas calculates the precompiled contract's base gas rate.
 func (p Precompile) RequiredGas(input []byte) uint64 {
+	if len(input) < 4 {
+		return 0
+	}
+
```
without adding this logic in the precompiles trying to send a transfer to this address would cause a panic instead of returning an error gracefully

GAtom22 replied:
can you please add this as a comment in the code?

GAtom22 commented on precompiles/vesting/vesting.go:
```diff
@@ -37,6 +40,10 @@ type Precompile struct {
 
 // RequiredGas returns the required bare minimum gas to execute the precompile.
 func (p Precompile) RequiredGas(input []byte) uint64 {
+	if len(input) < 4 {
```
ditto

0xstepit commented on app/app.go:
```diff
@@ -941,8 +950,21 @@ func (app *Evmos) BlockedAddrs() map[string]bool {
+	}
+
+	for _, precompile := range blockedPrecompilesHex {
+		blockedAddrs[sdk.AccAddress(ethcommon.HexToAddress(precompile).Bytes()).String()] = true
```
This seems to be a common chain of function calls, we could create a function like `EthHexToCosmosAddress`. What do you think?

0xstepit reviewed (approved):
Nice job @MalteHerrmann 💪
